pydantic/tools.py

```diff
@@ -35,15 +36,20 @@ def schema_of(
     schema_generator: type[GenerateJsonSchema] = GenerateJsonSchema,
 ) -> dict[str, Any]:
     """Generate a JSON schema (as dict) for the passed model or dynamically generated one"""
-    json_schema_generator = schema_generator(by_alias=by_alias, ref_template=ref_template)
-    if hasattr(type_, '__pydantic_core_schema__'):
-        core_schema = type_.__pydantic_core_schema__
-    else:
-        core_schema = _generate_schema.GenerateSchema(True, None).generate_schema(type_)
-    json_schema = json_schema_generator.generate(core_schema)
-    if title is not None:
-        json_schema['title'] = title
-    return json_schema
+    if not isinstance(title, (str, type(None))):  # pragma: no cover
+        warnings.warn(
+            'The type_name parameter is deprecated. parse_obj_as no longer creates temporary models',
+            DeprecationWarning,
+            stacklevel=2,
+        )
```

Comment:
This logic doesn't seem right to me — it's checking `title` to say something about `parse_obj_as` and temporary models.

I will also add that I think the `title` value is being mishandled here — I think we need to add something like `if callable(title): title = title()`

Reply:
The signature is `Callable[[Type[Any]], str]`. So does it expect the type itself? If so, what's the point of a callback if the type is also a parameter?

Reply:
```suggestion
        title = title(type_)
```

Is this really what we need? In abstract it seems like a pointless API, but if it's there for historic reasons it maybe makes sense...

Reply:
> If so, what's the point of a callback if the type is also a parameter?

I think at some point we thought it was a cleaner API to do it this way (I think I was the person who added the NameFactory, for `parse_obj_as` (which was originally `parse_as_type`). Probably it never needed to exist.

Reply:
Looks like it was my suggestion, and just came from analogy with the name generation stuff for generic models (which is done by overriding a method).  https://github.com/pydantic/pydantic/pull/934#discussion_r341298903 for reference.

I think you are right that we never needed to support name factories here, and it's a good idea to deprecate it.

Reply:
Thanks for digging. So the deprecation we have now should be okay?

Reply:
Your suggestion above with `title(type_)` is correct, if we just emit a deprecation warning rather than raising an error, I think we should call.